From fb192fd56408383a238b45fefa2edda94deec35e Mon Sep 17 00:00:00 2001
From: Xianzhong <b07117@freescale.com>
Date: Fri, 22 Feb 2013 00:23:53 +0800
Subject: [PATCH 06/30] ENGR00251253 Fixed UI flicker with swap rectangle

JB4.2 UI flicker in setting user name edit window.

The root cause is imx6 gpu 3d has alignment limitation
in resolving and cannot resolve front buffer correctly,
this is a workaround to draw the whole screen in 3d composition.

Signed-off-by: Xianzhong <b07117@freescale.com>
---
 services/surfaceflinger/SurfaceFlinger.cpp | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/services/surfaceflinger/SurfaceFlinger.cpp b/services/surfaceflinger/SurfaceFlinger.cpp
index c99212d..0641635 100644
--- a/services/surfaceflinger/SurfaceFlinger.cpp
+++ b/services/surfaceflinger/SurfaceFlinger.cpp
@@ -14,6 +14,8 @@
  * limitations under the License.
  */
 
+/* Copyright (C) 2013 Freescale Semiconductor, Inc. */
+
 #define ATRACE_TAG ATRACE_TAG_GRAPHICS
 
 #include <stdint.h>
@@ -1704,7 +1706,17 @@ void SurfaceFlinger::doDisplayComposition(const sp<const DisplayDevice>& hw,
         // we can redraw only what's dirty, but since SWAP_RECTANGLE only
         // takes a rectangle, we must make sure to update that whole
         // rectangle in that case
-        dirtyRegion.set(hw->swapRegion.bounds());
+        const int32_t id = hw->getHwcDisplayId();
+        HWComposer& hwc(getHwComposer());
+        HWComposer::LayerListIterator cur = hwc.begin(id);
+        const HWComposer::LayerListIterator end = hwc.end(id);
+
+        const bool hasGlesComposition = hwc.hasGlesComposition(id) || (cur==end);
+        if(hasGlesComposition) {//should draw the whole screen due to gpu3d limitation
+            dirtyRegion.set(hw->bounds());
+        }else {
+            dirtyRegion.set(hw->swapRegion.bounds());
+        }
     } else {
         if (flags & DisplayDevice::PARTIAL_UPDATES) {
             // We need to redraw the rectangle that will be updated
-- 
1.8.0

